spinalcordtoolbox/deepseg/monai.py

```diff
@@ -0,0 +1,212 @@
+                module.bias = nn.init.constant_(module.bias, 0)
+
+
+def prepare_data(path_image, path_out, crop_size=(64, 160, 320)):
+    # create a temporary datalist containing the image
+    # boiler plate keys to be defined in the MSD-style datalist
+    params = {
+        "description": "my-awesome-SC-image",
+        "labels": {"0": "background", "1": "soft-sc-seg"},
+        "modality": {"0": "MRI"},
+        "tensorImageSize": "3D",
+        "test": [{"image": path_image}]
+    }
+    dataset = os.path.join(path_out, "temp_msd_datalist.json")
+    with open(dataset, "w") as jsonFile:
+        jsonFile.write(json.dumps(params, indent=4, sort_keys=True))
+    test_files = load_decathlon_datalist(dataset, True, "test")
+    os.remove(dataset)
+
+    # define test transforms
+    transforms_test = inference_transforms_single_image(crop_size=crop_size)
+
+    # define post-processing transforms for testing; taken (with explanations) from
+    # https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/torch/unet_inference_dict.py#L66
+    test_post_pred = Compose([
+        EnsureTyped(keys=["pred"]),
+        Invertd(keys=["pred"], transform=transforms_test,
+                orig_keys=["image"],
+                meta_keys=["pred_meta_dict"],
+                nearest_interp=False, to_tensor=True),
+    ])
+    test_ds = CacheDataset(data=test_files, transform=transforms_test, cache_rate=0.75, num_workers=8)
+    test_loader = DataLoader(test_ds, batch_size=1, shuffle=False, num_workers=8, pin_memory=True)
+
+    return test_loader, test_post_pred
+
+
+def inference_transforms_single_image(crop_size):
+    return Compose([
+        LoadImaged(keys=["image"], image_only=False),
+        EnsureChannelFirstd(keys=["image"]),
+        Orientationd(keys=["image"], axcodes="RPI"),
+        Spacingd(keys=["image"], pixdim=(1.0, 1.0, 1.0), mode=2),
+        ResizeWithPadOrCropd(keys=["image"], spatial_size=crop_size, ),
+        DivisiblePadd(keys=["image"], k=2 ** 5),
+        # pad inputs to ensure divisibility by no. of layers nnUNet has (5)
+        NormalizeIntensityd(keys=["image"], nonzero=False, channel_wise=False),
+    ])
+
+
+def postprocessing(batch, test_post_pred):
+    # take only the highest resolution prediction
+    batch["pred"] = batch["pred"][0]
+
+    # NOTE: monai's models do not normalize the output, so we need to do it manually
+    if bool(F.relu(batch["pred"]).max()):
+        batch["pred"] = F.relu(batch["pred"]) / F.relu(batch["pred"]).max()
+    else:
+        batch["pred"] = F.relu(batch["pred"])
+
+    post_test_out = [test_post_pred(i) for i in decollate_batch(batch)]
+
+    pred = post_test_out[0]['pred'].cpu()
+
+    # clip the prediction between 0.5 and 1
+    # turns out this sets the background to 0.5 and the SC to 1 (which is not correct)
+    # details: https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/issues/71
+    pred = torch.clamp(pred, 0.5, 1)
```

Comment:
We are now investigating on which threshold to choose! probably won't be 0.5

Reply:
Ah! I realize now that the clamping/binarization is redundant, because `sct_deepseg` includes a `-thr` argument to perform binarizaiton depending on user preference.

So, I was thinking that on the SCT side of things, we could output the full softseg by default, then let the user choose the threshold themselves? (See: https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4345#discussion_r1462337603)

Reply:
Hmm, how about keeping the thresholded output as the default and letting user disable the thresholding if a soft output is required? 

We're noticing that while soft outputs have its advantages, users might not know how to interpret the soft values and what to do with them, etc. Whereas for the binary outputs, it is what's usually expected... 

What do you think? @sandrinebedard

Reply:
Added in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4345/commits/3a1ef65f77dc5bd5245332caa79f9b264403d5ec!